qiskit/pulse/logical_elements_frames.py

```diff
@@ -0,0 +1,499 @@
+        """Raise a ``PulseError`` if the qubit index is invalid. Namely, check if the index is a
+        non-negative integer.
+
+        Raises:
+            PulseError: If ``index`` is a negative integer.
+        """
+        pass
+        if not isinstance(index, (int, np.integer)) or index < 0:
+            raise PulseError("Qubit index must be a non-negative integer")
+
+
+class MixedFrame:
+    """Representation of a ``LogicalElement`` and ``Frame`` combination.
+
+    Every pulse needs to be associated with both a ``LogicalElement`` and a ``Frame``. The combination
+    of the two is called a mixed frame, and is represented by a ``MixedFrame`` object. The ``MixedFrame``
+    is closely related to the legacy ``Channel``, but provides a more flexible framework.
+
+    In most cases the ``MixedFrame`` is used more by the compiler, and a pulse program can be written
+    without ``MixedFrame``s, by setting ``LogicalElement`` and ``Frame`` independently. However,
+    in some cases using ``MixedFrame``s can better convey the meaning of the code, and change the
+    compilation process. One example are shift/set frequency/phase instructions which are not
+    broadcasted to other ``MixedFrame``s if applied on a specific ``MixedFrame`` (unlike the behavior
+    of ``Frame``). A second example is the ``CRMixedFrame`` which indicates to the compiler what is the
+    purpose of the pulses played on the mixed frame, and allows for extra validation - namely, validating
+    that qubits are coupled.
+    """
+
+    def __init__(self, logical_element: LogicalElement, frame: Frame):
+        """Create ``MixedFrame``.
+
+        Args:
+            logical_element: The logical element associated with the mixed frame.
+            frame: The frame associated with the mixed frame.
+        """
+        self._logical_element = logical_element
+        self._frame = frame
+        self._hash = hash((self._logical_element, self._frame))
+
+    @property
+    def logical_element(self) -> LogicalElement:
+        """Return the ``LogicalElement`` of this mixed frame."""
+        return self._logical_element
+
+    @property
+    def frame(self) -> Frame:
+        """Return the ``Frame`` of this mixed frame."""
+        return self._frame
+
+    @property
+    def name(self) -> str:
+        """Return the name of the mixed frame."""
+        return f"MixedFrame({self.logical_element},{self.frame})"
+
+    def __repr__(self) -> str:
+        return self.name
+
+    def __eq__(self, other: "MixedFrame") -> bool:
+        """Return True iff self and other are equal, specifically, iff they have the same logical
+        element and frame.
+
+        Args:
+            other: The mixed frame to compare to this one.
+
+        Returns:
+            True iff equal.
+        """
+        return self._logical_element == other._logical_element and self._frame == other._frame
+
+    def __hash__(self) -> int:
+        return self._hash
+
+
+class CRMixedFrame(MixedFrame):
```

Comment:
Can you write a separate PR for this class after we complete the rest of IR? I want to see the minimal code for the first development cycle.